Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -590,10 +592,38 @@ private class TopSitesDataSource: NSObject, UICollectionViewDataSource {
         self.historySites = historySites
     }
 
-    private func filterSuggestedSites() {
+    private func filterAndMergeSuggestedSites() {
+        // We need to filter from the complete suggested sites list every time.
+        // This is needed for edge cases, such as, clearing private data
+        // with a deduped topsite will not display the suggested site again
+        // unless you reload the Top Sites panel.
+        suggestedSites = SuggestedSites.asArray()
+
         for url in profile.prefs.arrayForKey("topSites.deletedSuggestedSites") as! [String] {
-            let domainURL = NSURL(string: url)?.normalizedHost() ?? url
-            suggestedSites = SuggestedSites.asArray().filter({ $0.tileURL.normalizedHost()?.URLString != domainURL.URLString })
+            let domainURL = extractDomainURL(url)
+            suggestedSites = suggestedSites.filter({ $0.tileURL.normalizedHost()?.URLString != domainURL.URLString })
+        }
+
+        var i = 0
+        while i < self.historySites.count {
+            let domainURL = extractDomainURL(self.historySites[i].url)
+            if suggestedSites.map({$0.tileURL.normalizedHost()!.URLString}).contains(domainURL) {
+                for site in suggestedSites {
+                    if site.tileURL.normalizedHost()?.URLString == domainURL.URLString {
+                        self.historySites[i] = site
+                    }
+                }
+            }
+            i++
+        }
+
+        for site in self.historySites {
```

Comment:
Don't do this here...do this as part of the step above, since you're already looking at both arrays to compare the sites. Basically:

1. Call `map` on `historySites`, setting the result back to `historySites`.
2. For each site, get its domain.
3. Use `if let index = suggestedSites.indexOf(domain)` to see if the site's domain is a suggested site.
4. If so, get a reference to the suggested site using the index, and also remove it from the suggested sites using that index. Then return that suggested site.
5. Otherwise, simply return the site itself.